src/array.cr

```diff
@@ -581,22 +581,54 @@ class Array(T)
 
   # Returns all elements that are within the given range.
   #
-  # Negative indices count backward from the end of the array (-1 is the last
-  # element). Additionally, an empty array is returned when the starting index
-  # for an element range is at the end of the array.
-  #
-  # Raises `IndexError` if the range's start is out of range.
+  # The first element in the returned array is `self[range.begin]` followed
+  # by the next elements up to index `range.end` (or `range.end - 1` if the range
+  # is exclusive).
+  # If there are fewer elements in `self`, the returned array is shorter than
+  # `range.size`.
   #
   # ```
   # a = ["a", "b", "c", "d", "e"]
   # a[1..3]    # => ["b", "c", "d"]
-  # a[4..7]    # => ["e"]
+  # # range.end > array.size
+  # a[3..7]    # => ["d", "e"]
+  # ```
+  #
+  # Open ended ranges are clamped at the start and end of the array, respectively.
+  #
+  # ```
+  # # open ended ranges
+  # a[2..]     # => ["c", "d", "e"]
+  # a[..2]     # => ["a", "b", "c"]
+  # ```
+  #
+  # Negative range values are added to `self.size`, thus they are treated as
+  # indices counting from the end of the array, `-1` designating the last element.
+  #
+  # ```
+  # # negative indices, both ranges are equivalent for `a`
+  # a[1..3]     # => ["b", "c", "d"]
+  # a[-4..-2]   # => ["b", "c", "d"]
+  # # Mixing negative and positive indices, both ranges are equivalent for `a`
+  # a[1..-2]    # => ["b", "c", "d"]
+  # a[-4..3]    # => ["b", "c", "d"]
+  # ```
+  #
+  # Raises `IndexError` if the start index it out of range (`range.begin >
+  # self.size || range.begin < -self.size). If `range.begin == self.size` an
+  # empty array is returned. If `range.begin > range.end`, an empty array is
+  # returned.
+  #
+  # ```
+  # # range.begin > array.size
   # a[6..10]   # raise IndexError
+  # # range.begin == array.size
   # a[5..10]   # => []
-  # a[-2...-1] # => ["d"]
-  # a[2..]     # => ["c", "d", "e"]
+  # # range.begin > range.end
+  # a[3..1]    # => []
+  # a[-2..-4]  # => []
```

Comment:
I think it would be more clarifying to add the mixed positive/negative index examples:

```
a[-2..1]    # => []
a[3..-4]    # => []
```
so that the user is reminded that `range.begin > range.end` is not just the value of the range, but instead the `begin` and `end` relative to the Array.

If the user is still agnostically thinking of `range.begin > range.end`, they'll realize that `-2 > 1` is false, and thus they might recall that range.begin and range.end is actually relative to the Array